Sources/STARSDK/Cryptography/IdentityManager.swift

```diff
@@ -0,0 +1,163 @@
+//  Created by Loïc Gardiol on 30.03.20.
+
+import Foundation
+import CommonCrypto
+
+
+/// Responsible for generating the secret keys (SKs) and deriving the Ephemeral IDs (EphID)
+/// Currently not thread-safe.
+class IdentityManager {
+    
+    static let shared = IdentityManager()
+    
+    init() {}
+
+    /// Last nbStoredEpochs SKs, in chronological order. Last is current.
+    /// Generated on first access, then persisted securely and rotated.
+    /// Guarenteed to be non-empty
+    private var SKs: [SK] {
+        let persistenceKey = "SKs"
+        var sks: [SK] = {
+            if let persisted: [SK] = SecurePersistence.shared.object(key: persistenceKey) {
+                return persisted
+            }
+            #warning("TODO If Keychain access fails (device just rebooted for example) and we can't read the previously persisted keys, we are going to override them here.")
+            let initialSK = SK(epoch: Epoch.current, keyData: Utils.generateRandomKey(nbBits: GlobalParameters.SKNbBits))
+            return [initialSK]
+        }()
+        let currentEpoch = Epoch.current
+        if let last = sks.last, last.epoch != currentEpoch {
+            var sk = last
+            while sk.epoch < currentEpoch {
+                sk = sk.next
+                sks.append(sk)
+            }
+        }
+        if sks.count > GlobalParameters.nbStoredEpochs {
+            sks = Array(sks.dropFirst(sks.count - Int(GlobalParameters.nbStoredEpochs)))
+        }
+        SecurePersistence.shared.set(sks, key: persistenceKey)
+        return sks
+    }
+    
+    private var currentEphIDCache: EphID?
+    
+    var currentEphID: EphID {
+        let currentSK = self.SKs.last! // doc says so
+        if let cached = self.currentEphIDCache, cached.epoch == currentSK.epoch {
+            return cached
+        }
+        let broadcast = "broadcast".data(using: .utf8)!
+        let hmac = Utils.hmacSHA256(msg: broadcast, key: currentSK.keyData)
+        let truncated = hmac.dropLast(hmac.count-GlobalParameters.EphIDNbBytes).withUnsafeBytes { Data($0) }
+        let ephId = EphID(epoch: currentSK.epoch, data: truncated)
+        self.currentEphIDCache = ephId
+        return ephId
+    }
+}
+
+struct SK: Codable, CustomStringConvertible {
+    let epoch: Epoch
+    let keyData: Data
+    
+    var next: SK {
+        let data = Utils.sha256(data: self.keyData)
+        return SK(epoch: self.epoch.next, keyData: data)
+    }
+    
+    var description: String {
+        return "<SK_\(self.epoch.index): \(self.keyData.hexEncodedString)>"
+    }
+}
+
+private class Utils {
+    /// Generates a random key using CommonCrypto.
+    /// Assumes nbBits is a multiple of 8.
+    static func generateRandomKey(nbBits: Int) -> Data {
```

Comment:
Is there a particular reason why we want bits and not bytes? It seems we can anyways only generate a multiple of 8bits. if we really want bits, we should use floor for the number of bytes generated. Otherwise we might get OutOfBounds (e.g. we want 9bits but there are only 8bits generated)

Reply:
Went with bits because keys are typically expressed in bits, I'm totally fine to move to bytes for the reason you mention.